File: LeetCode/236. Lowest Common Ancestor of a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        class Helper {
            TreeNode lca(TreeNode node) {
                return lowestCommonAncestor(node, p, q);
            }
        }
        Helper helper = new Helper();
        if (root == null || root == p || root == q) return root;
        TreeNode left = helper.lca(root.left);
        TreeNode right = helper.lca(root.right);

        if (left == null) return right;
        if (right == null) return left;
        return root;
    }
}

File: LeetCode/236. Lowest Common Ancestor of a Binary Tree.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        if (root == null || root == p || root == q) return root
        val left = lowestCommonAncestor(root.left, p, q)
        val right = lowestCommonAncestor(root.right, p, q)
        return if (left == null) right
               else if (right == null) left
               else root
    }
    //if the subtree has no p or q, it will eventually return null
    //when the recursion is back to the parent, pick the non-null child as result
    //but if both children are not null, return the parent node
}
